contracts/zero-ex/contracts/src/bridges/mixins/MixinUniswap.sol

```diff
@@ -0,0 +1,197 @@
+    function tokenToTokenTransferInput(
+        uint256 tokensSold,
+        uint256 minTokensBought,
+        uint256 minEthBought,
+        uint256 deadline,
+        address recipient,
+        address toTokenAddress
+    )
+        external
+        returns (uint256 tokensBought);
+}
+
+contract MixinUniswap {
+
+    using LibERC20TokenV06 for IERC20TokenV06;
+
+    /// @dev Mainnet address of the WETH contract.
+    IEtherTokenV06 constant private WETH = IEtherTokenV06(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
+    /// @dev Mainnet address of the `UniswapExchangeFactory` contract.
+    IUniswapExchangeFactory constant private UNISWAP_EXCHANGE_FACTORY = IUniswapExchangeFactory(0xc0a47dFe034B400B47bDaD5FecDa2621de6c4d95);
+
+    function _tradeUniswap(
+        address toTokenAddress,
+        uint256 sellAmount,
+        bytes memory bridgeData
+    )
+        internal
+        returns (uint256 boughtAmount)
+    {
+        // Decode the bridge data to get the `fromTokenAddress`.
+        (address fromTokenAddress) = abi.decode(bridgeData, (address));
+
+        // Get the exchange for the token pair.
+        IUniswapExchange exchange = _getUniswapExchangeForTokenPair(
+            fromTokenAddress,
+            toTokenAddress
+        );
+
+        // Convert from WETH to a token.
+        if (fromTokenAddress == address(WETH)) {
+            // Unwrap the WETH.
+            WETH.withdraw(sellAmount);
+            // Buy as much of `toTokenAddress` token with ETH as possible
+            boughtAmount = exchange.ethToTokenTransferInput{ value: sellAmount }(
+                // Minimum buy amount.
+                1,
+                // Expires after this block.
+                block.timestamp,
+                // Recipient is `this`.
+                address(this)
+            );
+
+        // Convert from a token to WETH.
+        } else if (toTokenAddress == address(WETH)) {
+            // Grant the exchange an allowance.
+            IERC20TokenV06(fromTokenAddress).approveIfBelow(
+                address(exchange),
+                sellAmount
+            );
+            // Buy as much ETH with `fromTokenAddress` token as possible.
+            boughtAmount = exchange.tokenToEthSwapInput(
+                // Sell all tokens we hold.
+                sellAmount,
+                // Minimum buy amount.
+                1,
+                // Expires after this block.
+                block.timestamp
+            );
+            // Wrap the ETH.
+            WETH.deposit{ value: boughtAmount }();
+        // Convert from one token to another.
+        } else {
+            // Grant the exchange an allowance.
+            IERC20TokenV06(fromTokenAddress).approveIfBelow(
+                address(exchange),
+                sellAmount
+            );
+            // Buy as much `toTokenAddress` token with `fromTokenAddress` token
+            boughtAmount = exchange.tokenToTokenTransferInput(
```

Comment:
Nit: We should use [`tokenToTokenSwapInput()`](https://github.com/Uniswap/uniswap-v1/blob/master/contracts/uniswap_exchange.vy#L292). One less param to abi-encode.